src/storage-client/src/types/connections.rs

```diff
@@ -532,27 +511,56 @@ impl CsrConnection {
-            // downstream incorrectly starts using this port.
-            const DUMMY_PORT: u16 = 11111;
-
-            // TODO: use types to enforce that the URL has a string hostname.
-            let host = self
-                .url
-                .host_str()
-                .ok_or_else(|| anyhow!("url missing host"))?;
-            let privatelink_host = mz_cloud_resources::vpc_endpoint_name(aws_privatelink_id);
-            let addrs: Vec<_> = net::lookup_host((privatelink_host, DUMMY_PORT))
-                .await
-                .context("resolving PrivateLink host")?
-                .collect();
-            client_config = client_config.resolve_to_addrs(host, &addrs)
+        let mut tokens = vec![];
+        match &self.tunnel {
+            Tunnel::Direct => {}
+            ssh_tunnel @ Tunnel::Ssh { .. } => {
+                // TODO: use types to enforce that the URL has a string hostname.
+                let host = self
+                    .url
+                    .host_str()
+                    .ok_or_else(|| anyhow!("url missing host"))?;
+
+                let (local_port, token) = ssh_tunnel
+                    .build_ssh_tunnel_for_url(
+                        secrets_reader,
+                        host,
+                        // Default to the default http port, but this
+                        // could default to 8081...
+                        self.url.port().unwrap_or(80),
+                        "csr",
+                    )
+                    .await?;
+
+                client_config = client_config
+                    .override_url(Url::parse(&format!("http://localhost:{}", local_port)).unwrap());
+                tokens.push(token);
+            }
+            Tunnel::AwsPrivatelink(connection) => {
+                assert!(connection.port.is_none());
+
+                // `net::lookup_host` requires a port but the port will be ignored
+                // when passed to `resolve_to_addrs`. We use a dummy port that will
+                // be easy to spot in the logs to make it obvious if some component
+                // downstream incorrectly starts using this port.
+                const DUMMY_PORT: u16 = 11111;
+
+                // TODO: use types to enforce that the URL has a string hostname.
+                let host = self
+                    .url
+                    .host_str()
+                    .ok_or_else(|| anyhow!("url missing host"))?;
+                let privatelink_host =
+                    mz_cloud_resources::vpc_endpoint_name(connection.connection_id);
+                let addrs: Vec<_> = net::lookup_host((privatelink_host, DUMMY_PORT))
+                    .await
+                    .context("resolving PrivateLink host")?
+                    .collect();
+                client_config = client_config.resolve_to_addrs(host, &addrs)
+            }
         }
 
-        client_config.build()
+        client_config.build_with_tokens(tokens)
```

Comment:
As mentioned above, I think we should instead expose a wrapper in this crate

```rust
struct CsrClient {
    inner: mz_ccsr::Client,
    tokens: Box<dyn Any>,
}

impl Deref for CsrClient {
    type Target = mz_ccsr::Client;
}
```

so that we don't have to pollute the `mz_ccsr` crate with the tokens required for SSH tunneling.

Reply:
this seems fine to me!

Reply:
this infects the `interchange` crate, requiring that all the types a generic over `C: Deref<mz_ccsr::Client>`, as opposed to just holding a `mz_ccsr::Client`, does that seem reasonable?

Reply:
Ah, yeah! That seems ok. Alternatively we can invent a crate that lives above `mz_ccsr` but below `mz_interchange` to put the `TokenHoldingCsrClient`.

Reply:
It seems kinda weird that `mz_interchange` depends on `mz_storage_client`. Weird.

Reply:
Oh, I see why. It's an easy fix.

Reply:
@benesch the pr now reflects the `: Deref` bounds in interchange, which looks fine for now, now that I have done it!